runtime/doc/lsp.txt

```diff
@@ -800,6 +794,58 @@ set_log_level({level})                               *vim.lsp.set_log_level()*
                 See also: ~
                     |vim.lsp.log_levels|
 
+start({config}, {opts})                                      *vim.lsp.start()*
+                Create a new LSP client and start a language server or reuses
+                an already running client if one is found matching `name` and
+                `root_dir`. Attaches the current buffer to the client.
+
+                Example:
+>
+
+    vim.lsp.start({
+       name = 'my-server-name',
+       cmd = {'name-of-language-server-executable'},
+       root_dir = vim.fs.dirname(vim.fs.find({'pyproject.toml', 'setup.py'})[1]),
+    })
+<
+
+                See |lsp.start_client| for all available options. The most
+                important are:
+
+                `name` is an arbitrary name for the LSP client. It should be
+                unique per language server.
+
+                `cmd` the command as list - used to start the language server. The
+                command must be present in the `$PATH` environment variable or an absolute path to the executable.
+                Shell constructs like `~` are NOT expanded.
```

Comment:
I don't recall all the details, but I wonder if it'd be worthwhile documenting how Windows handles (or rather doesn't handle) certain executables [through libuv](https://github.com/libuv/libuv/blob/2a1b880f5439e074ef4d5c556f626b1044fb6781/src/win/process.c#L304-L308). A large portion of servers are distributed as `.cmd`, `.ps1` or `.bat` scripts on Windows, which would not play nicely here as libuv wouldn't recognize them as executable. This is currently handled by lspconfig by using `cmd.exe /C` as an intermediary for launching server processes.

I think this could potentially be a source of high volume issues for Windows users.

Reply:
I think that's out of scope of this PR

Reply:
I was just thinking if this is a new API that is intended to be used directly by users it might lead to some hard-to-troubleshoot issues. I'd wager a vast majority of users would expect ergonomics similar to their shell environment, with path expansion and `PATHEXT` to be respected.

Just an example from the perspective of Windows users, the following would not function:

```lua
vim.lsp.start {
  cmd = { "typescript-language-server", "--stdio" },
  ..
}
```

Assuming you have your npm PATH set up correctly; jumping into a shell and typing `where typescript-language-server` would look fine, going into Neovim and trying `:echo executable("typescript-language-server")` would [also look fine](https://github.com/neovim/neovim/blob/c632f64e247c672e425f609bb47a9ab0517a4c31/src/nvim/os/fs.c#L323). I think this has been discussed before, but I think spawning RPC processes via a shell should be considered as more LSP functionality finds its way into core (or some other alternative to meet user expectations). There might be some other future plans that I'm not aware of so apologies if I'm overreaching

Reply:
It is a new API that is intended to be used by moderately advanced users. New users (especially on Windows) will likely still rely on a lspconfig or lspinstaller plugin (which hopefully can be simplified using this new API).

Supporting "shell ergonomics for everyone" is _far_ outside the scope of this PR. (Although it may be reasonable to improve `vim.fs` to make things somewhat easier.)